### PR TITLE
Rewrite delimcpy to use memchr and Copy, not per-byte

### PR DESCRIPTION
**Pull request by khwilliamson**
Prior to this commit delimcpy() parsed its input byte-by-byte, looking
for a particular character, and copies the input to the output stopping
just before the first such occurrence.

memchr() is much faster for finding a single character.

The complication is that if the character is preceded by a backslash,
it doesn't count as that character, it is considered to be escaped, and
parsing continues to the first unescaped occurrence, if any.  Each
escaping backslash is not copied.

The new routine looks for the character with memchr, sees if it is
escaped.  If not, Copy does the whole copy at once.  If it is escaped,
it uses Copy up to that backslash, and repeats the process.

I suspect that it is a bug that the routine did not consider the
possibility of the escape itself being escaped, so not counting as
escaping the searched for character.  But I have retained that behavior.

### CODE REVIEW

**Comment by khwilliamson** [comment]
Does anyone have a feeling if it is a bug that it has never counted backslashes before the delimiter?

**Comment by hvds** [comment]
> The complication is that if the character is preceded by a backslash,
> it doesn't count as that character, it is considered to be escaped, and
> parsing continues to the first unescaped occurrence, if any. Each
> escaping backslash is not copied.
[...]
> I suspect that it is a bug that the routine did not consider the
> possibility of the escape itself being escaped, so not counting as
> escaping the searched for character. But I have retained that behavior.

I don't agree that's what's happening - if allow_escapes is ~~true~~ false, backlashes are not treated as special at all (courtesy ba0a4150, which sadly did not change the docs); if ~~false~~ true, when it sees a backslash not followed by the delimiter, the backslash is copied and pointers advanced, and then it falls through to the rest, which copies the following character. So if `x` is the delimiter, I'd expect `\x\\\x\\x` to copy `x\\x\\` and then see the delimiter and stop.

Hugo

[Edited to correct boolean sense of allow_escapes]

**Comment by hvds** [comment]
Note that 19e16554 shows how one can write tests for this behaviour.

**Comment by khwilliamson** [comment]
There are two functions, one for ignoring the delimiter; one to notice it.  It used to be that they called a third function that did all the work with a flag as to which behavior to follow.  That is what @hvds is referring to with 'allow_escapes'.  But I rewrote the first function to just be a simple memchr and a Copy.  That leaves what this branch is doing, to rewrite the second function to not be so convoluted.  In this branch that flag is removed, and is effectively true; to account for escapes

I agree with @hvds that it should work like his example, but in fact blead does not.  With input `\x\\\x\\x` instead of returning `x\\x\\` blead returns `x\\\\`, which is what I would have expected it to given my claim of the bug.

I can make my replacement work like Hugo and I say it should.  But I'll await feedback.

In the meantime, this branch currently has my fix reverted, and the beginnings of an API test for delimcpy added.  If someone wants to play with it, you can easily try out test cases; otherwise I think gdb would be required

**Comment by hvds** [comment]
> I agree with @hvds that it should work like his example, but in fact blead does not. With input `\x\\\x\\x` instead of returning `x\\x\\` blead returns `x\\\\`, which is what I would have expected it to given my claim of the bug.

Your testcase is being corrupted by interpolation of '\\'.

Trying this:
```
my $b = "\\";
my $x = "x";
@ret = test_delimcpy("$b$x$b$b$b$x$b$b$x", 'x', 100);
```
gives:
```
$VAR1 = [
          [
            'x\\\\x\\\\',  # note each backslash is being shown here as \\
            6,             # but the length confirms it is correct
            8
          ]
        ];

```

Hugo

File: ext/XS-APItest/t/delimcpy.t [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#!perl -w
+use strict;
+
+use Test::More;
+use XS::APItest;
+
+
+my @ret;
+@ret = test_delimcpy('\x\\\x\\x', 'x', 100);
```
Comment (hvds) [inline ext/XS-APItest/t/delimcpy.t:9]: This perl code will suffer from single-quotish interpretation of backslashes.

**Comment by khwilliamson** [comment]
I've now examined this further, and have some more clarity.  My concern does in fact lie with @hvds example`\x\\\x\\x`
I'm concerned with the asymmetry that when copying the backslashes for the first two x'es there it strips off one in each sequence.  But when copying the two backslashes for the final 'x', it leaves them both in.  So it strips an x for odd numbers of backslashes and doesn't for even numbers.  That asymmetry seems wrong to me.  The test suite passes whichever behavior is done.

**Comment by hvds** [comment]
> I've now examined this further, and have some more clarity. My concern does in fact lie with @hvds example`\x\\\x\\x`
> I'm concerned with the asymmetry that when copying the backslashes for the first two x'es there it strips off one in each sequence. But when copying the two backslashes for the final 'x', it leaves them both in. So it strips an x for odd numbers of backslashes and doesn't for even numbers. That asymmetry seems wrong to me. The test suite passes whichever behavior is done.

That is precisely the intent - it strips the escape from the delimiter, but passes through all other escapes. So `\x\y\zx` will pass through `x\y\zx`. From the docs added by @iabyn in 19e16554:
```
   * With allow_escape, converts \<delimiter> to <delimiter>, while leaves
   * \<non-delimiter> as-is.
```

**Comment by khwilliamson** [comment]
I can take your word for it that this is the intent, but that comment 

> converts \\delimiter to delimiter, while leaves \\non-delimiter as-is.

doesn't do it for me.  I just don't understand how it should be obvious that `\\delimiter` isn't a form of `\delimiter`, without the further explanation that you have to count the preceding parentheses to determine if a character is the delimiter or not.

What was obvious to me was that the code was sloppy, not realizing, for example, that one of the two main cases was actually trivial, and didn't need a loop at all, much less two nested ones.  That lowered my confidence that the authors had even considered this edge case that was glaring to me.  I have found plenty of such instances in the code base; and have created some myself where I was blind to what was plain as day to someone else.

**Comment by hvds** [comment]
> I can take your word for it that this is the intent, but that comment
> 
> > converts \delimiter to delimiter, while leaves \non-delimiter as-is.
> 
> doesn't do it for me. I just don't understand how it should be obvious that `\\delimiter` isn't a form of `\delimiter`, without the further explanation that you have to count the preceding parentheses to determine if a character is the delimiter or not.

Sorry, I've little idea what you're saying here. It is substantially confused by the fact that in github I can see your quoted text `... converts \delimiter ...` with a single backslash, while in the email copy it showed up as a double backslash. The original from the source code that I was trying to share certainly has a single backslash.

So I don't understand what you mean by "how it should be obvious that `\\delimiter`" [double-backslash] "isn't a form of `\delimiter`" [single-backslash]. And I'm further confused by "counting parentheses", unless that's a typo for backslashes.

> What was obvious to me was that the code was sloppy, not realizing, for example, that one of the two main cases was actually trivial, and didn't need a loop at all, much less two nested ones.

I don't understand this either - I don't see two nested loops anywhere, and a loop is definitely necessary, even if it could in some cases be supplied by a library function. I guess you're talking about the change introduced by ba0a4150f6, but in the context that change still looks reasonable to me (other than the shortage of docs) - the fact that the no_escape variant could be optimized by reimplementing it as a completely independent function does not seem so obvious or so important an improvement that I'd consider the lack of it a signal about the quality of the code.

But I guess all that is less relevant than: what do you want to do with this PR?

**Comment by khwilliamson** [comment]
parentheses instead of backslashes.  I shouldn't stay up late to write something technical.

And it does come down to a PR.  I would like to update this PR to one that behaves 'properly' for some proper definition of 'properly'.  To that end I wrote some tests, starting with the one @hvds pointed out my fallacies in.   These include edge cases.  I have applied these tests to the current implementation, and have found one worrisome (to me) issue.  A mandatory parameter is &retlen.  The value returned in it is aspirational rather than real.  It returns the size the copy would be if there is enough room in the destination, not the size that actually gets stored.  If the destination is big enough to hold the entire copy, there is no difference, but if it isn't, adding this length to the buffer beginning, as the code in core does, and assuming that is where to write the next things will lead to an out of bounds write.  I think the value returned should be the actual size written.

**Comment by hvds** [comment]
> A mandatory parameter is &retlen. The value returned in it is aspirational rather than real. It returns the size the copy would be if there is enough room in the destination, not the size that actually gets stored. If the destination is big enough to hold the entire copy, there is no difference, but if it isn't, adding this length to the buffer beginning, as the code in core does, and assuming that is where to write the next things will lead to an out of bounds write. I think the value returned should be the actual size written.

Right, that one does look tricky. Looking briefly at the two uses in toke.c, the first appears to blindly accept it as the actual length copied while the second takes advantage of the actual returned value to raise an error if we would have gone out of bounds.

It seems easy to fix that first call to correct its use of the currently returned value. It seems rather harder to fix the second one if you change what it returns - it would no longer be indicating anything to show that the buffer was too small, other than the hint that the buffer was exactly filled. So my first instinct would be to retain the current return value (but actually document it), and fix that first caller.

**Comment by khwilliamson** [comment]
On 8/24/20 5:56 AM, Hugo van der Sanden wrote:
>     A mandatory parameter is &retlen. The value returned in it is
>     aspirational rather than real. It returns the size the copy would be
>     if there is enough room in the destination, not the size that
>     actually gets stored. If the destination is big enough to hold the
>     entire copy, there is no difference, but if it isn't, adding this
>     length to the buffer beginning, as the code in core does, and
>     assuming that is where to write the next things will lead to an out
>     of bounds write. I think the value returned should be the actual
>     size written.
> 
> Right, that one does look tricky. Looking briefly at the two uses in 
> toke.c, the first appears to blindly accept it as the actual length 
> copied while the second takes advantage of the actual returned value to 
> raise an error if we would have gone out of bounds.
> 
> It seems easy to fix that first call to correct its use of the currently 
> returned value. It seems rather harder to fix the second one if you 
> change what it returns - it would no longer be indicating anything to 
> show that the buffer was too small, other than the hint that the buffer 
> was exactly filled. So my first instinct would be to retain the current 
> return value (but actually document it), and fix that first caller.
> 

delimcpy is an API function, though I don't see actual uses in CPAN

I think it would be better to return ((STRLEN) -1) if the buffer would 
have overflowed.

This is what we do in many other parts of the API.

This shouldn't change the behavior of code that checks that the return 
is in bounds.

And it would cause a likely segfault for code that blindly continues on. 
  A segfault is preferable to an attack vector.

**Comment by hvds** [comment]
> I think it would be better to return ((STRLEN) -1) if the buffer would have overflowed.

Ah ok, that makes sense, it might even work. (I assume we're still talking about what should be stored in retlen, rather than the actual return value of the function.)

**Comment by khwilliamson** [comment]
On 8/24/20 9:28 AM, Hugo van der Sanden wrote:
>     I think it would be better to return ((STRLEN) -1) if the buffer
>     would have overflowed.
> 
> Ah ok, that makes sense, it might even work. (I assume we're still 
> talking about what should be stored in retlen, rather than the actual 
> return value of the function.)

Right.  The function return value is sane, and documented properly, so I 
wouldn't want to change that.

**Comment by khwilliamson** [comment]
This was superceded by #18198
